src/rgw/rgw_role.cc

```diff
@@ -0,0 +1,220 @@
+#include <errno.h>
+
+#include <boost/date_time/posix_time/posix_time.hpp>
+
+#include "common/errno.h"
+#include "common/Formatter.h"
+#include "common/ceph_json.h"
+#include "rgw_rados.h"
+
+#include "include/types.h"
+#include "rgw_string.h"
+
+#include "rgw_common.h"
+#include "rgw_tools.h"
+#include "rgw_role.h"
+
+#define dout_subsys ceph_subsys_rgw
+
+using namespace std;
+
+int RGWRole::store_info(bool exclusive)
+{
+  bufferlist bl;
+  ::encode(*this, bl);
+  return rgw_put_system_obj(store, store->get_zone_params().roles_pool, id,
+                bl.c_str(), bl.length(), exclusive, NULL, real_time(), NULL);
+}
+
+int RGWRole::store_name(bool exclusive)
+{
+  RGWNameToId nameToId;
+  nameToId.obj_id = id;
+
+  bufferlist bl;
+  ::encode(nameToId, bl);
+  return rgw_put_system_obj(store, store->get_zone_params().roles_pool, name,
+              bl.c_str(), bl.length(), exclusive, NULL, real_time(), NULL);
+}
+
+int RGWRole::create(bool exclusive)
+{
+  int ret;
+
+  /* check to see the name is not used */
+  ret = read_id(name, id);
+  if (exclusive && ret == 0) {
+    ldout(cct, 0) << "ERROR: name " << name << " already in use for role id "
+                    << id << dendl;
+    return -EEXIST;
+  } else if ( ret < 0 && ret != -ENOENT) {
+    ldout(cct, 0) << "failed reading role id  " << id << ": "
+                  << cpp_strerror(-ret) << dendl;
+    return ret;
+  }
+
+  /* create unique id */
+  uuid_d new_uuid;
+  char uuid_str[37];
+  new_uuid.generate_random();
+  new_uuid.print(uuid_str);
+  id = uuid_str;
+
+  // Creation time
+  creation_date = boost::posix_time::to_iso_extended_string(
+                          boost::posix_time::microsec_clock::universal_time());
```

Comment:
does the standard library support this string format? i'd prefer to use `ceph::real_clock` instead of boost if at all possible